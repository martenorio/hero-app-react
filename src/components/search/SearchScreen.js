import { useMemo } from 'react'
import { useNavigate, useLocation } from "react-router-dom";
import queryString from 'query-string'
import { useForm } from "../../hooks/useForm";
import { getHeroesByName } from "../../helpers/getHeroesByName";
import { HeroCard } from "../hero/HeroCard";

export const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const { q = ''} = queryString.parse(location.search)

  const [formValues, handleInputChange ] = useForm({ searchText : q });
  const { searchText } = formValues;

  const heroesFiltered = useMemo( () => getHeroesByName(q), [q]);

  const handleSearch = (event) => {
    // console.log(searchText);
    event.preventDefault();
    navigate(`?q=${ searchText }`);
  }


  return (
      <>
        <h1>Búsquedas</h1>
        <hr />
        <div className="row">
          <div className="col-5">
            <h4>Formulario</h4>
            <hr />
            <form onSubmit={ handleSearch }>
              <input
              type="text" 
              placeholder="Busca un héroe"
              className="form-control" 
              name="searchText"
              autoComplete="off"
              onChange={ handleInputChange }
              value={searchText}
              />
              <div className="d-grid gap-2">
                <button 
                type="submit" className="btn btn-primary mt-2 btn-block">
                  Buscar
                </button>

              </div>
            </form>
          </div>
          <div className="col-7">
            <h4>Resultados</h4>
            <hr />
            {
              (q === '') ?
              <div className="alert alert-info animate__animated animate__fadeIn"> Buscar un héroe </div>
              : (heroesFiltered.length === 0) && <div className="alert alert-danger animate__animated animate__fadeIn"> No hay resultados : {q} </div>
            }
            {
              heroesFiltered.map( hero => (
                <HeroCard key={hero.id} {...hero}></HeroCard>
              ))
            }
          </div>
        </div>
      </>
  )
}
