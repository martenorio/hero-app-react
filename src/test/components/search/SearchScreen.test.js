import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

const mockNavigate = jest.fn();

//El mock es necesario, por que dentro de la función 
//se ejecuta navigate de  useNavigate que forma parte de 
//react-router-dom y para que la prueba 
//no tenga errores con la funcion de navigate se simula la ejecución
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}))

describe('Search Screen tests', () => {
    
    test('Should display correctly default values', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Buscar un héroe')
    })
    test('should display batman and set queryString values on input', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>
        )
        expect(wrapper.find('input').prop('value')).toBe('batman');
    })
    test('should show an error if the hero is not found', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchScreen />
            </MemoryRouter>
        );
        expect(wrapper.find('.alert-danger').text().trim()).toBe('No hay resultados : batman123');
    })
    test('should execute navigate to new screen', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        );
        wrapper.find('input').simulate('change',{ 
            target:{ 
                name:'searchText',
                value:'batman'
            }
        });
        wrapper.find('form').prop('onSubmit')({
            preventDefault:()=>{} 
        });
        expect( mockNavigate ).toHaveBeenCalled();
        expect( mockNavigate ).toHaveBeenCalledWith('?q=batman')
    });
})