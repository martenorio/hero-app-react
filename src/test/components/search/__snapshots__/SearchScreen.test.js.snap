// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Screen tests Should display correctly default values 1`] = `
Array [
  <h1>
    Búsquedas
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Formulario
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Busca un héroe"
          type="text"
          value=""
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn btn-primary mt-2 btn-block"
            type="submit"
          >
            Buscar
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info animate__animated animate__fadeIn"
      >
         Buscar un héroe 
      </div>
    </div>
  </div>,
]
`;
